src/core/server/ui_settings/ui_settings_client.ts

```diff
@@ -209,10 +212,20 @@ export class UiSettingsClient implements IUiSettingsClient {
     }
   }
 
-  private async read({
-    ignore401Errors = false,
-    autoCreateOrUpgradeIfMissing = true,
-  }: ReadOptions = {}): Promise<Record<string, any>> {
+  private async read() {
+    const cachedValue = this.cache.get();
+    if (cachedValue) {
+      return cachedValue;
+    }
+
+    const value = this.onReadHook(await this._read());
+    this.cache.set(value);
+    return value;
+  }
+
+  private async _read({ autoCreateOrUpgradeIfMissing = true }: ReadOptions = {}): Promise<
```

Comment:
Another question for my info only: Both `read` and `_read` are private. Is the use of the underscore in `_read` to indicate the method is _very_ private?

Reply:
I will inline `read` function into `getUserProvided`.